### ISSUE TITLE
Support for runtime context object?

### ISSUE THREAD

**Issue by cleverplatypus** [pegjs/pegjs#666]
### Issue type

- **Bug Report:** 
- **Feature Request:** 
- **Question:** _yes_
- **Not an issue:** 

### Prerequisites
- **Can you reproduce the issue?:** _yes_ 
- **Did you search the repository issues?:** _yes_ 
- **Did you check the forums?:** _yes_ 
- **Did you perform a web search (google, yahoo, etc)?:** _yes_ 

### Description
Is there any reason that context data can only be set at compile time?
I see many situations where the compiled grammar would need to access a context that changes at runtime.
Compiling the grammar for every use case doesn't seem to be a viable solution.

Unless I'm missing something....

**Example code:**

I'm using PEG for a decorator in ExpressJS that controls access to endpoints:
```javascript
@Auth('role(admin) || adgroup(ACME_Developer)')
myEndpointMethod() {
}
```
The grammar relies on a `profile` variable that is supposed to contain the current user's profile containing roles and adgroups.
The profile is only available when the endpoint is hit.
I'd expect the compiled grammar to work something like this:
```javascript
const isAuthorised = myAuthGrammar.parse('role(admin) || adgroup(ACME_Developer)', {
    profile : {
          roles : ['tester'],
          adgroups : ['ACME_Tester']
    }
});
```

### Software
- **PEG.js:**

**Comment by hildjj** [comment]
This might be a good place for a `& {predicate}` block.  Parse a token with something like `$[a-z_]i+`, then check it in a list of valid tokens in the predicate.  Here is an intentionally-simplified example (no paren grouping) that parses your inputs and rejects invalid roles/adgroups:

```js
'use strict'

const pegjs = require('pegjs')

const src = `
authorized
  = perm:perm rest:(_ o:op _ p:perm { return [o, p]})* { rest.unshift(perm); return rest }

perm = pn:permname _ "(" _ n:name & { return options[pn].includes(n) } _")" {
  return { [pn]: n }
}

permname
  = "role"i { return "roles" }
  / "adgroup"i { return "adgroups" }

name = $[a-z_]i+

op
  = $"||"
  / $"&&"

_ = $[ \t]*
`
const parser = pegjs.generate(src)

for (const a of process.argv.slice(2)) {
  console.log(parser.parse(a, {
    roles: ['admin'],
    adgroups: ['ACME_Developer']
  }))
}
```

**Comment by cleverplatypus** [comment]
Thanks @hildjj.

Are you saying that the `options` object is available in the grammar and it contains whatever I pass in as a second parameter of parser.parse()?

That's not clear from the docs. Is it a "hack" or is it going to be working in future versions of PEG?

I ended up creating a grammar that outputs and AST that then is interpreted by my code.

```javascript
{
   "any": [
      {
         "all": [
            {
               "adgroup": "ACME_Developer"
            },
            {
               "role": "admin"
            }
         ]
      },
      {
         "adgroup": "supergod"
      }
   ]
}
```

**Comment by hildjj** [comment]
Yes, the docs are a little unclear about this, but it's supported behavior. I'll create a bug on peggyjs/peggy to improve the docs.
